### PR TITLE
Add a simple key<->Container persistent store.

### PR DESCRIPTION
**Pull request by aluzzardi**
Signed-off-by: Andrea Luzzardi <aluzzardi@gmail.com>

### CODE REVIEW

File: state/store.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	if err := s.restore(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Store) path(key string) string {
```
Comment (tnachen) [inline state/store.go]: I haven't seen where this is used yet, so ignore me if this is unrelevant.
Should the key be santizied (lowercased, trim, etc?)

**Comment by aluzzardi** [comment]
Yeah, sorry about that.

It will be used by this PR: https://github.com/docker/swarm/pull/236

File: state/store.go (hunk 1 above)
Comment (vieux) [inline state/store.go]: Key will be IDs so there is no need to cleanup. @aluzzardi can you confirm ?

Comment (aluzzardi) [inline state/store.go]: Indeed !

Comment (tnachen) [inline state/store.go]: Thanks for clarifying!

File: state/store.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+		// Load the object back.
+		value, err := s.load(path.Join(s.RootDir, file))
+		if err != nil {
+			return err
+		}
+
+		// Extract the key.
+		key := file[0 : len(file)-len(extension)]
+		if len(key) == 0 {
```
Comment (tnachen) [inline state/store.go]: I know this is a very specific store for swarm, so trying to focus on swarm specific use cases.
But I think if we're going to be checking for the key length here, there is nothing stopping the Add() to pass a empty key. If somehow that happened, on restore we simply can't initialize anymore. I think it's worth fixing on the Add side.

File: state/store.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		file := fileinfo.Name()
+
+		// Verify the file extension.
+		extension := filepath.Ext(file)
+		if extension != ".json" {
+			return fmt.Errorf("invalid file extension for filename %s (%s)", file, extension)
+		}
+
+		// Load the object back.
+		value, err := s.load(path.Join(s.RootDir, file))
```
Comment (tnachen) [inline state/store.go]: Should we stop the restore if we can't load a single file?
I wonder if there are situations a crash before writing the whole file can cause half writes to the disk, and then cause us never able to restore. 
I would consider log error but still able to use the cache.

File: state/store.go (hunk 2 above)
Comment (aluzzardi) [inline state/store.go]: Done. PTAL

**Comment by tnachen** [comment]
LGTM :)

**Comment by aluzzardi** [comment]
@vieux ?

**Comment by vieux** [comment]
LGTM

**Pull request merged**